Add tests for global function aliases/clones

markstory opened the pull request:
- Add tests for some of the namespaced functions that didn't have test coverage yet.
- Add test coverage for global functions. I haven't done the i18n functions yet as I wanted to get alignment on the approach we want to take here first.
- Fix a bug in deprecationWarning that would indicate the incorrect line number.
- Added a functions.php to match what we have in 5.x. This will allow users to opt-in to the behavior they want to have now and have less work to do in a future upgrade.

Refs #17101

markstory committed:
Add tests for global function aliases/clones

- Add tests for some of the namespaced functions that didn't have
  test coverage yet.
- Add test coverage for global functions. I haven't done the i18n functions yet
  as I wanted to get alignment on the approach we want to take here first.
- Fix a bug in deprecationWarning that would indicate the incorrect line
  number.
- Added a functions.php to match what we have in 5.x. This will allow
  users to opt-in to the behavior they want to have now and have less
  work to do in a future upgrade.

ADmad commented:
> Added a functions.php to match what we have in 5.x. This will allow users to opt-in to the behavior they want to have now and have less work to do in a future upgrade.

For BC each functions.php already loads it's corresponding functions_global.php, so is this really necessary?

Are you adding this just so that we can update the 4.x skeleton too and have the same include as we added for 5.x?

ADmad commented on tests/TestCase/Core/GlobalFunctionsTest.php:
```diff
@@ -0,0 +1,371 @@
+use Cake\TestSuite\TestCase;
+use stdClass;
+
+require_once CAKE . 'Core/functions_global.php';
```
Including this file means if one of these functions gets used without namespace, in files which are loaded after this test case, we won't we able to detect it.

Perhaps using the `@runTestsInSeparateProcesses` for the test class can avoid that?

markstory commented:
> Are you adding this just so that we can update the 4.x skeleton too and have the same include as we added for 5.x?

Yes. It won't do anything useful in 4.x but it will let people 'upgrade' early.

ADmad commented on src/functions.php:
```diff
@@ -0,0 +1,24 @@
+require __DIR__ . '/Collection/functions_global.php';
+require __DIR__ . '/Core/functions_global.php';
+require __DIR__ . '/I18n/functions_global.php';
+require __DIR__ . '/Routing/functions_global.php';
```
`Error/functions_global.php` is missing.

markstory replied:
Thanks!

markstory committed:
Fix failing tests and add captureError()

I was bothered by the warnings in our test output and thought this was
a reasonable way to solve some of them.

markstory committed:
Fix compatibility issues.

markstory committed:
Remove bad tests.

Now that the deprecation lines are correct they aren't the same and
message duplication can't be tested.

markstory committed:
Fix captureError()

Also fix static analyzers

markstory commented:
I think I got the tests fixed (at least they pass locally) and I added a new method so we can progress on the PHPUnit deprecations we (and likely application code) are facing. The `captureError()` method provides a simple API that makes writing assertions simple. If folks like it I can update the balance of our error level warnings with it.

ADmad commented:
I think you have missed my comment regarding running tests which load the global functions in a separate process.

markstory commented:
> I think you have missed my comment regarding running tests which load the global functions in a separate process.

I tried running global functions in separate tests and it went poorly as PHPUnit encountered a bunch of errors. I can try moving the global function tests to a separate suite.

markstory committed:
Run global functions separately and include coverage.

othercorey commented:
Looks good.
